src/basic/fileio.c

```diff
@@ -310,6 +310,100 @@ int verify_file(const char *fn, const char *blob, bool accept_extra_nl) {
+         * the first read are buffered for the second read. The other sort uses "raw"
+         * reads which always go direct to the device. In the latter case, the content of
+         * the virtual file must be retrieved with a single read otherwise a second read
+         * might get the new value instead of finding EOF immediately. That's the reason
+         * why the usage of fread(3) is prohibited in this case as it always performs a
+         * second call to read(2) looking for EOF. See issue 13585. */
+
+        fd = open(filename, O_RDONLY|O_CLOEXEC);
+        if (fd < 0)
+                return -errno;
+
+        /* Start size for files in /proc which usually report a file size of 0. */
+        size = LINE_MAX;
+
+        for (;;) {
+                if (fstat(fd, &st) < 0)
+                        return -errno;
+
+                if (!S_ISREG(st.st_mode))
+                        return -EBADF;
+
+                /* Be prepared for files from /proc which generally report a file size of 0. */
+                if (st.st_size > 0)
+                        size = st.st_size;
+
+                if (size > READ_FULL_BYTES_MAX)
+                        return -E2BIG;
+
+                p = realloc(buf, size + 1);
+                if (!p)
+                        return -ENOMEM;
+                buf = TAKE_PTR(p);
+
+                for (;;) {
+                        /* Read one more byte so we can detect if the content of the file
+                         * has already changed. */
+                        n = read(fd, buf, size + 1);
+                        if (n >= 0)
+                                break;
+
+                        if (errno != -EINTR)
+                                return -errno;
+                }
+
+                /* Consider a short read as EOF */
+                if (n <= size)
+                        break;
+
+                /* Hmm... either we read too few bytes from a file in /proc or less
+                 * likely the content of the file might have been changed (and is now
+                 * bigger) while we were processing, let's try again either with a bigger
+                 * guessed size or the new file size. */
+
+                if (lseek(fd, 0, SEEK_SET) < 0)
+                        return -errno;
+
+                size = size * 2;
+        }
+
+        if (n < size) {
+                p = realloc(buf, n + 1);
+                if (!p)
+                        return -ENOMEM;
+                buf = TAKE_PTR(p);
+        }
+
+        if (!ret_size) {
+                /* Safety check: if the caller doesn't want to know the size of what we
+                 * just read it will rely on the trailing NUL byte. But if there's an
+                 * embedded NUL byte, then we should refuse operation as otherwise
+                 * there'd be ambiguity about what we just read. */
+
+                if (memchr(buf, 0, n))
```

Comment:
This doesn't look right.

You're always allocating size + 1, and later in this function's normal return path the \0 is stored in that +1 of space.

Why isn't this instead just checking for ret_size before storing the size, and always going through the normal return path of ensuring there's a \0 terminator?  Don't look for an existing terminator in the stream, just always add one.

Reply:
As the comment explained, if the caller doesn't want to know the number of bytes returned by this function *and* it contains NUL bytes (besides the trailing one) then there's no way for the caller to know how long the returned buf is so the returned contents is unusable.

Reply:
BTW this part was taken from read_full_stream_full()

Reply:
I think the comment doesn't make sense.

If the caller didn't ask for a size, and the contents contain a null byte, the contents are self null-terminated.  It's not unusable, it just contains a null byte, and that fulfills all safety requirements for a C string which is the only likely sane intention of use when no size was requested.  Why are we prohibiting reading an already null-terminated string when the caller doesn't request the size?

I admit I'm not exactly focused on systemd/computer stuff at the moment, but this strikes me as broken.

IMNSHO the size should just be stored if requested, and a potentially redundant null terminator always added in the space allocated for it.  I'm all for defensive programming but this doesn't strike me as a meaningful example of it.

What am I missing?

Reply:
At least if the caller expects to read a string (and therefore didn't ask for a size), and the function returned more bytes after the first NUL bytes, then it's likely that  something wrong is happening.

Maybe we could just warn...

@poettering ?